### PR TITLE
Avoid class attribute `_keep_in_fp32_modules` being modified

### LINKED ISSUE #25910: Load `instructblip-flan-t5-xl` changes the results of `instructblip-vicuna-7b`
When loading `instructblip-flan-t5-xl` using `InstructBlipForConditionalGeneration`, it will change the (logits) output of `instructblip-vicuna-7b`. See the reproduction section.

If I tried to load `facebook/opt-350m` (using `OPTForCausalLM`), it won't cause this situation (i.e. `instructblip-vicuna-7b` have the same logits).

If I load `instructblip-flan-t5-xl`, it won't change the results of `facebook/opt-350m` (loaded in 8-bit).

Something is strange here and requires further investigation.

### Reproduction

The following one shows `Salesforce/instructblip-vicuna-7b` is affected by `instructblip-flan-t5-xl`

```
import requests
import torch
from PIL import Image
from transformers import InstructBlipProcessor, InstructBlipForConditionalGeneration, OPTForCausalLM

torch_device = "cuda"

url = "https://raw.githubusercontent.com/salesforce/LAVIS/main/docs/_static/Confusing-Pictures.jpg"
image = Image.open(requests.get(url, stream=True).raw).convert("RGB")
prompt = "What is unusual about this image?"

processor = InstructBlipProcessor.from_pretrained("Salesforce/instructblip-vicuna-7b")
inputs = processor(images=image, text=prompt, return_tensors="pt").to(torch_device, torch.float16)

def check_vicuna_logits(load_flan_t5=False, load_opt=False):

    if load_flan_t5:
        _ = InstructBlipForConditionalGeneration.from_pretrained(
           "Salesforce/instructblip-flan-t5-xl",
           # don't matter `float16` or `bfloat16`
           torch_dtype=torch.bfloat16,
        )

    if load_opt:
        _ = OPTForCausalLM.from_pretrained(
           "facebook/opt-350m",
           # don't matter `float16` or `bfloat16`
           torch_dtype=torch.bfloat16,
        )

    model = InstructBlipForConditionalGeneration.from_pretrained(
        "Salesforce/instructblip-vicuna-7b", load_in_8bit=True
    )
    with torch.no_grad():
        logits = model(**inputs).logits

    print(logits[0, :3, :3])


# [ -3.4902, -12.5078,   8.4141]
check_vicuna_logits(load_flan_t5=False)

# [ -3.4902, -12.5078,   8.4141]
check_vicuna_logits(load_flan_t5=False, load_opt=True)

# Once `Salesforce/instructblip-flan-t5-xl` is loaded (even just once), `logits` changes to `-3.4609`.
# [ -3.4609, -12.0156,   8.3750]
check_vicuna_logits(load_flan_t5=True)

# continue to be `-3.4609`
# [ -3.4609, -12.0156,   8.3750]
check_vicuna_logits(load_flan_t5=False)
```

The following one shows `Opt` won't be affected by `/instructblip-flan-t5-xl`

```python
import torch

from transformers import InstructBlipForConditionalGeneration, OPTForCausalLM, AutoTokenizer

torch_device = "cuda"
tokenizer = AutoTokenizer.from_pretrained("facebook/opt-350m")

prompt = "Hey, are you conscious? Can you talk to me?"
inputs = tokenizer(prompt, return_tensors="pt")   # .to(torch_device, torch.float16)

for k, v in inputs.items():
    if torch.is_floating_point(v):
        inputs[k] = v.to(device=torch_device, dtype=torch.bfloat16)

def check_opt_logits(load_flan_t5=False):

    if load_flan_t5:
        _ = InstructBlipForConditionalGeneration.from_pretrained(
           "Salesforce/instructblip-flan-t5-xl",
           # don't matter `float16` or `bfloat16`
           torch_dtype=torch.bfloat16,
        )

    model = OPTForCausalLM.from_pretrained(
       "facebook/opt-350m",
       load_in_8bit=True,
    )

    with torch.no_grad():
        logits = model(**inputs).logits

    print(logits[0, :3, :3])

check_opt_logits(load_flan_t5=False)
check_opt_logits(load_flan_t5=True)
```

### Expected behavior

The outputs of `Salesforce/instructblip-vicuna-7b` **SHOULD NOT** be affected by if we load `instructblip-flan-t5-xl` or not.

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

Fix #25910

Currently, for `InstructBlip` which can combine/use different architecture as component, we have the following block

https://github.com/huggingface/transformers/blob/abd253103478b21faafb7c9a6e7a1a7d1effe757/src/transformers/models/instructblip/modeling_instructblip.py#L1280-L1281

in order to extent the class attribute `_keep_in_fp32_modules` of `InstructBlipPreTrainedModel`. This modification of class attribute causes some issue:
- loading `Salesforce/instructblip-flan-t5-xl` changes `_keep_in_fp32_modules` from `[]` to `["wo"]`, because the language model is a `T5PreTrainedModel` which has `_keep_in_fp32_modules = ["wo"]`.
  - Then loading `Salesforce/instructblip-vicuna-7b` will keep `qformer.embeddings.word_embeddings.weight` in `fp32`, due to a bug before the fix #26589.
- If we don't load `flan-t5` but just `vicuna`, `qformer.embeddings.word_embeddings.weight` will be in `fp16`.
and the weight values are slightly different -> outputs are different -> CI failure.

**Modifying a class attribute is bad** in general, and we have failing test. This CI tries to incorporate components' in `_keep_in_fp32_modules` without modifying the `_keep_in_fp32_modules` in the parent model class.

### CODE REVIEW

File: src/transformers/modeling_utils.py
Comment (ydshieh) [inline src/transformers/modeling_utils.py:1148]: Taking instance specific fp32 module info. account (which may contain components' `_keep_in_fp32_modules`.

File: src/transformers/models/instructblip/modeling_instructblip.py
Comment (ydshieh) [inline src/transformers/models/instructblip/modeling_instructblip.py:1281]: we should NOT modify class attribute `_keep_in_fp32_modules`!

File: src/transformers/models/instructblip/modeling_instructblip.py
Comment (ydshieh) [inline src/transformers/models/instructblip/modeling_instructblip.py:1281]: use `_keep_in_fp32_modules_in_instance` to store the information from the component. This instance attribute will be used in `_keep_in_fp32_full_modules` to get the full list of fp32 module list.

**Review by younesbelkada — APPROVED** [review]
Merci @ydshieh ! Can you make sure that the CI I have introduced in https://github.com/huggingface/transformers/pull/26589 pass by merging your PR with transformers main ? 🙏

**Comment by ydshieh** [comment]
nothing, just didn't get an approval from core maintainers. Will ping one.

**Comment by ydshieh** [comment]
Before merge this PR (if being approved), I have to apply the change to some recently modified files.

File: src/transformers/modeling_utils.py
Comment (ArthurZucker) [inline src/transformers/modeling_utils.py:1147]: no our usual docstring format

**Review by ArthurZucker — COMMENTED** [review]
Thanks! If `_keep_in_fp32_modules` is just an instance attribute instead of a class attribute it might help / remove the need to have extra complexity imo

**Comment by ArthurZucker** [comment]
No you can put it in BerPretraneModel's __init__

**Comment by ydshieh** [comment]
Yes, that works well, thanks.

**Pull request closed without merging**
